File: 2-NORMAL/Python/5-bfs-dfs-graph-traversal/proposed-solution/graph.py
from collections import deque

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.adj = [[] for _ in range(vertices)]

    def add_edge(self, u, v):
        self.adj[u].append(v)
        self.adj[v].append(u)  # Assuming undirected graph

    def bfs(self, start):
        visited = [False] * self.V
        queue = deque([start])
        visited[start] = True
        result = []

        while queue:
            vertex = queue.popleft()
            result.append(vertex)
            for neighbor in self.adj[vertex]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)
        return result

    def dfs_util(self, v, visited, result):
        visited[v] = True
        result.append(v)
        for neighbor in self.adj[v]:
            if not visited[neighbor]:
                self.dfs_util(neighbor, visited, result)

    def dfs(self, start):
        visited = [False] * self.V
        result = []
        self.dfs_util(start, visited, result)
        return result

    def bfs_disconnected(self):
        visited = [False] * self.V
        result = []
        for i in range(self.V):
            if not visited[i]:
                queue = deque([i])
                visited[i] = True
                while queue:
                    vertex = queue.popleft()
                    result.append(vertex)
                    for neighbor in self.adj[vertex]:
                        if not visited[neighbor]:
                            visited[neighbor] = True
                            queue.append(neighbor)
        return result

    def dfs_disconnected(self):
        visited = [False] * self.V
        result = []
        for i in range(self.V):
            if not visited[i]:
                self.dfs_util(i, visited, result)
        return result

File: 2-NORMAL/JavaScript/5-bfs-dfs-graph-traversal/proposed-solution/graph.js
// graph.js - Graph class with BFS and DFS traversals

class Graph {
  constructor(vertices) {
    this.V = vertices;
    this.adj = Array.from({ length: vertices }, () => []);
  }

  addEdge(u, v) {
    this.adj[u].push(v);
    this.adj[v].push(u); // Undirected graph
  }

  bfs(start) {
    const visited = new Array(this.V).fill(false);
    const queue = [];
    const result = [];

    visited[start] = true;
    queue.push(start);

    while (queue.length > 0) {
      const vertex = queue.shift();
      result.push(vertex);

      for (const neighbor of this.adj[vertex]) {
        if (!visited[neighbor]) {
          visited[neighbor] = true;
          queue.push(neighbor);
        }
      }
    }
    return result;
  }

  dfsUtil(v, visited, result) {
    visited[v] = true;
    result.push(v);

    for (const neighbor of this.adj[v]) {
      if (!visited[neighbor]) {
        this.dfsUtil(neighbor, visited, result);
      }
    }
  }

  dfs(start) {
    const visited = new Array(this.V).fill(false);
    const result = [];
    this.dfsUtil(start, visited, result);
    return result;
  }

  bfsDisconnected() {
    const visited = new Array(this.V).fill(false);
    const result = [];

    for (let i = 0; i < this.V; i++) {
      if (!visited[i]) {
        const queue = [];
        visited[i] = true;
        queue.push(i);

        while (queue.length > 0) {
          const vertex = queue.shift();
          result.push(vertex);

          for (const neighbor of this.adj[vertex]) {
            if (!visited[neighbor]) {
              visited[neighbor] = true;
              queue.push(neighbor);
            }
          }
        }
      }
    }
    return result;
  }

  dfsDisconnected() {
    const visited = new Array(this.V).fill(false);
    const result = [];

    for (let i = 0; i < this.V; i++) {
      if (!visited[i]) {
        this.dfsUtil(i, visited, result);
      }
    }
    return result;
  }
}

export default Graph;
